src/Components/Web.JS/src/Services/NavigationManager.ts

```diff
@@ -116,18 +153,97 @@ function performInternalNavigation(absoluteInternalHref: string, interceptedLink
   resetScrollAfterNextBatch();
 
   if (!replace) {
-    history.pushState(state, /* ignored title */ '', absoluteInternalHref);
+    currentHistoryIndex++;
+    history.pushState({
+      userState: state,
+      _index: currentHistoryIndex,
+    }, /* ignored title */ '', absoluteInternalHref);
   } else {
-    history.replaceState(state, /* ignored title */ '', absoluteInternalHref);
+    history.replaceState({
+      userState: state,
+      _index: currentHistoryIndex,
+    }, /* ignored title */ '', absoluteInternalHref);
+  }
+
+  await notifyLocationChanged(interceptedLink);
+}
+
+function navigateHistoryWithoutPopStateCallback(delta: number): Promise<void> {
+  return new Promise(resolve => {
+    const oldPopStateCallback = popStateCallback;
+
+    popStateCallback = () => {
+      popStateCallback = oldPopStateCallback;
+      resolve();
+    };
+
+    history.go(delta);
```

Comment:
Could you clarify why we do this thing of using the event handler itself to temporarily remove and put back the event handler?

A possibly-simpler alternative would be just setting a flag saying "ignore popstate callbacks" before we call `history.go` and then clearing the flag after. Or is there some case where one navigation synchronously triggers another or something, causing us to need multiple levels of flag?

Reply:
My idea behind doing it this way was to avoid situations where some part of the code assumes that the navigation has completed and the `location` and `history` objects are up-to-date (because `history.go()` is asynchronous). For example, we probably want to wait until the navigation completes before calling `notifyLocationChanged`: https://github.com/dotnet/aspnetcore/blob/33653224a3000cfa94dde308919024228262f652/src/Components/Web.JS/src/Services/NavigationManager.ts#L222-L239
There are of course other ways to achieve this, but I thought making a helper enabling `async`/`await` would be easier to read and work with than having additional global state/callbacks keeping track of where we are in the navigation. However, I'm definitely open to considering alternative approaches that reduce the complexity of all this!